lib/src/Uncertainty/Algorithm/MetaModel/TensorApproximation/CanonicalTensorEvaluation.cxx

```diff
@@ -0,0 +1,243 @@
+                                                     const Indices & nk,
+                                                     const UnsignedInteger rank)
+: NumericalMathEvaluationImplementation()
+, functionFamilies_(functionFamilies)
+, degrees_(nk)
+, coefficients_(nk.getSize())
+, basis_(nk.getSize())
+{
+  const UnsignedInteger dimension = functionFamilies.getSize();
+  if (dimension != nk.getSize())
+    throw InvalidArgumentException(HERE) << "The number of function factories (" << dimension << ") is different from number of basis sizes (" << nk.getSize() << ")";
+  if (rank == 0)
+    throw InvalidArgumentException(HERE) << "Rank cannot be null";
+
+  for (UnsignedInteger j = 0; j < dimension; ++ j)
+  {
+    coefficients_[j] = NumericalSample(rank, nk[j]);
+    basis_[j] = Basis(nk[j]);
+    for (UnsignedInteger k = 0; k < nk[j]; ++ k)
+    {
+      basis_[j][k] = NumericalMathFunction(UniVariateFunctionEvaluation(functionFamilies[j].build(k)));
+    }
+  }
+}
+
+/* Virtual constructor */
+CanonicalTensorEvaluation * CanonicalTensorEvaluation::clone() const
+{
+  return new CanonicalTensorEvaluation(*this);
+}
+
+
+
+CanonicalTensorEvaluation::FunctionFamilyCollection CanonicalTensorEvaluation::getFunctionFamilies() const
+{
+  return functionFamilies_;
+}
+
+
+Indices CanonicalTensorEvaluation::getDegrees() const
+{
+  return degrees_;
+}
+
+
+/* Coefficients accessor along i-th component */
+void CanonicalTensorEvaluation::setCoefficients(const UnsignedInteger i, const UnsignedInteger j, const NumericalPoint & coefficients)
+{
+  coefficients_[j][i] = coefficients;
+}
+
+
+NumericalPoint CanonicalTensorEvaluation::getCoefficients(const UnsignedInteger i, const UnsignedInteger j) const
+{
+  return coefficients_[j][i];
+}
+
+
+/* Basis accessor along i-th component */
+Basis CanonicalTensorEvaluation::getBasis(const UnsignedInteger i) const
+{
+  return basis_[i];
+}
+
+
+void CanonicalTensorEvaluation::setRank(const UnsignedInteger rank)
+{
+  if (rank == 0)
+    throw InvalidArgumentException(HERE) << "Rank cannot be null";
+
+  const UnsignedInteger oldRank = getRank();
+
+  // initialize added tensor coefficients
+  for (UnsignedInteger j = 0; j < getInputDimension(); ++ j)
+  {
+    if (rank > oldRank)
+    {
+      // add coefficients
+      coefficients_[j].add(NumericalSample(rank - oldRank, degrees_[j]));
+    }
+    else if (rank < oldRank)
+    {
+      // remove coefficients
+      coefficients_[j] = NumericalSample(coefficients_[j], 0, rank);
+    }
+  }
+}
+
+
+UnsignedInteger CanonicalTensorEvaluation::getRank() const
+{
+  return coefficients_[0].getSize();
+}
+
+
+CanonicalTensorEvaluation CanonicalTensorEvaluation::getMarginalRank(const UnsignedInteger i) const
+{
+  CanonicalTensorEvaluation rankOne(getFunctionFamilies(), getDegrees());
+  for (UnsignedInteger j = 0; j < getInputDimension(); ++ j)
+  {
+    rankOne.setCoefficients(0, j, getCoefficients(i, j));
+  }
+  return rankOne;
+}
+
```

Comment:
I do not understand why you store `functionFamilies_`, it is already embedded inside `basis_`.  If I am not mistaken, you could write
```
  CanonicalTensorEvaluation rankOne(*this);
  rankOne.setRank(1);
  for (UnsignedInteger j = 0; j < getInputDimension(); ++ j)
  {
    rankOne.setCoefficients(0, j, getCoefficients(i, j));
  }
  return rankOne;
```
and then get rid of functionFamilies_ and getFunctionFamilies().